test/python/circuit/test_library.py

```diff
@@ -50,3 +55,108 @@ def test_inner_product(self):
         expected.cz(1, 4)
         expected.cz(2, 5)
         self.assertEqual(circuit, expected)
+
+
+@ddt
+class TestBasisChanges(QiskitTestCase):
+    """Test the basis changes."""
+
+    def to_matrix(self, circuit):
+        """Get the unitary matrix from simulating the circuit with the unitary simulator."""
+        backend = BasicAer.get_backend('unitary_simulator')
+        return execute(circuit, backend).result().get_unitary()
+
+    def assertQFTIsCorrect(self, qft, num_qubits=None, inverse=False, prepend_swaps=False):
+        """Assert that the QFT circuit produces the correct matrix.
+
+        Can be provided with an explicit number of qubits, if None is provided the number
+        of qubits is set to ``qft.num_qubits``.
+        """
+        if prepend_swaps:
+            circuit = QuantumCircuit(*qft.qregs)
+            for i in range(circuit.num_qubits // 2):
+                circuit.swap(i, circuit.num_qubits - i - 1)
+
+            qft = circuit + qft
+
+        simulated = self.to_matrix(qft)
+
+        num_qubits = num_qubits or qft.num_qubits
+        expected = np.empty((2 ** num_qubits, 2 ** num_qubits), dtype=complex)
+        for i in range(2 ** num_qubits):
+            for j in range(i, 2 ** num_qubits):
+                entry = np.exp(2 * np.pi * 1j * i * j / 2 ** num_qubits) / 2 ** (num_qubits / 2)
+                expected[i, j] = entry
+                if i != j:
+                    expected[j, i] = entry
+
+        if inverse:
+            expected = np.conj(expected)
+
+        np.testing.assert_array_almost_equal(simulated, expected)
+
+    @data(True, False)
+    def test_qft_matrix(self, inverse):
+        """Test the matrix representation of the QFT."""
+        num_qubits = 4
+        qft = QFT(num_qubits)
+        if inverse:
+            qft = qft.inverse()
+        self.assertQFTIsCorrect(qft, inverse=inverse)
+
+    def test_qft_mutability(self):
+        """Test the mutability of the QFT circuit."""
+        qft = QFT()
+
+        with self.subTest(msg='empty initialization'):
+            self.assertEqual(qft.num_qubits, 0)
+            self.assertEqual(qft.data, [])
+
+        with self.subTest(msg='changing number of qubits'):
+            qft.num_qubits = 3
+            self.assertQFTIsCorrect(qft, num_qubits=3)
+
+        with self.subTest(msg='test diminishing the number of qubits'):
+            qft.num_qubits = 1
+            self.assertQFTIsCorrect(qft, num_qubits=1)
+
+        with self.subTest(msg='test with swaps'):
+            qft.num_qubits = 4
+            qft.do_swaps = False
+            self.assertQFTIsCorrect(qft, prepend_swaps=True)
+
+        with self.subTest(msg='set approximation'):
+            qft.approximation_degree = 2
+            qft.do_swaps = True
+            with self.assertRaises(AssertionError):
+                self.assertQFTIsCorrect(qft)
```

Comment:
instead of checking that they are not equal (which could also happen if say there is a bug), can you instead check that they are equal with a looser equality threshold (`atol`)?

Reply:
Yeah I'm not entirely sure how the approximation degree affects the transformation matrix, setting atol even to `1e-1` still fails the test